backends/magma/ceed-magma-basis.c

```diff
@@ -0,0 +1,266 @@
+	    				     impl->dinterp1d, impl->dgrad1d, tmode, 
+   					     u + dim_ctr * u_dimstride + u_compstride * comp_ctr, u_elstride, 
+					     v + dim_ctr * v_dimstride + v_compstride * comp_ctr, v_elstride, 
+					     nelem, dim_ctr);       
+        }
+     }
+   }
+  if (emode & CEED_EVAL_WEIGHT) {
+    if (tmode == CEED_TRANSPOSE)
+      return CeedError(ceed, 1,
+                       "CEED_EVAL_WEIGHT incompatible with CEED_TRANSPOSE");
+    CeedInt Q = Q1d;
+    int eldofssize = CeedIntPow(Q, dim);
+    magmablas_dbasis_apply_batched_eval_weight(Q, dim, impl->dqweight1d, 
+					       v, eldofssize, 
+					       nelem);
+  }
+
+  if(emode!=CEED_EVAL_WEIGHT) {
+    ierr = CeedVectorRestoreArrayRead(U, &u); CeedChk(ierr);
+  }
+  ierr = CeedVectorRestoreArray(V, &v); CeedChk(ierr);
+  return 0;
+}
+
+#ifdef __cplusplus
+extern "C"
+#endif
+int CeedBasisDestroy_Magma(CeedBasis basis) 
+{
+  int ierr;
+  CeedBasis_Magma *impl;
+  ierr = CeedBasisGetData(basis, (void *)&impl); CeedChk(ierr);
+  
+  ierr = magma_free(impl->dqref1d); CeedChk(ierr);
+  ierr = magma_free(impl->dinterp1d); CeedChk(ierr);
+  ierr = magma_free(impl->dgrad1d); CeedChk(ierr);
+  ierr = magma_free(impl->dqweight1d); CeedChk(ierr);
+
+  ierr = CeedFree(&impl); CeedChk(ierr);
+
+  return 0;
+}
+
+#ifdef __cplusplus
+extern "C"
+#endif
+int CeedBasisCreateTensorH1_Magma(CeedInt dim, CeedInt P1d,
+    CeedInt Q1d, const CeedScalar *interp1d,
+    const CeedScalar *grad1d,
+    const CeedScalar *qref1d,
+    const CeedScalar *qweight1d,
+    CeedBasis basis) 
+{
+  int ierr;
+  CeedBasis_Magma *impl;
+  Ceed ceed;
+  ierr = CeedBasisGetCeed(basis, &ceed); CeedChk(ierr);
+  
+  ierr = CeedSetBackendFunction(ceed, "Basis", basis, "Apply",
+                                CeedBasisApply_Magma); CeedChk(ierr);
+  ierr = CeedSetBackendFunction(ceed, "Basis", basis, "Destroy",
+                                CeedBasisDestroy_Magma); CeedChk(ierr);
+
+  ierr = CeedCalloc(1,&impl); CeedChk(ierr);
+  ierr = CeedBasisSetData(basis, (void *)&impl); CeedChk(ierr);
+
+  // Copy qref1d to the GPU
+  ierr = magma_malloc((void**)&impl->dqref1d, Q1d*sizeof(qref1d[0]));
+  CeedChk(ierr);
+  magma_setvector(Q1d, sizeof(qref1d[0]), qref1d, 1, impl->dqref1d, 1);
+  
+  // Copy interp1d to the GPU
+  ierr = magma_malloc((void**)&impl->dinterp1d, Q1d*P1d*sizeof(interp1d[0]));
+  CeedChk(ierr);
+  magma_setvector(Q1d*P1d, sizeof(interp1d[0]), interp1d, 1, impl->dinterp1d, 1);
+
+  // Copy grad1d to the GPU
+  ierr = magma_malloc((void**)&impl->dgrad1d, Q1d*P1d*sizeof(grad1d[0]));
+  CeedChk(ierr);
+  magma_setvector(Q1d*P1d, sizeof(grad1d[0]), grad1d, 1, impl->dgrad1d, 1);
+
+  // Copy qweight1d to the GPU
+  ierr = magma_malloc((void**)&impl->dqweight1d, Q1d*sizeof(qweight1d[0]));
+  CeedChk(ierr);
+  magma_setvector(Q1d, sizeof(qweight1d[0]), qweight1d, 1, impl->dqweight1d, 1);
+
+  return 0;
+}
+
+#ifdef __cplusplus
+extern "C"
+#endif
```

Comment:
You should be able to use `CEED_INTERN` here.

Reply:
Changed in de205e78161a90a71fb703c7b1873f1fe194cf3c